runtime/sema/check_interface_declaration.go

```diff
@@ -357,62 +372,93 @@ func (checker *Checker) declareInterfaceMembers(declaration *ast.InterfaceDeclar
+		}
 
-	initializers := declaration.Members.Initializers()
-	interfaceType.InitializerParameters = checker.initializerParameters(initializers)
-	interfaceType.InitializerPurity = checker.initializerPurity(compositeKind, initializers)
+		// NOTE: determine initializer parameter types while nested types are in scope,
+		// and after declaring nested types as the initializer may use nested type in parameters
 
-	// Declare nested declarations' members
+		initializers := declaration.Members.Initializers()
+		interfaceType.InitializerParameters = checker.initializerParameters(initializers)
+		interfaceType.InitializerPurity = checker.initializerPurity(compositeKind, initializers)
 
-	for _, nestedInterfaceDeclaration := range declaration.Members.Interfaces() {
-		// resolve conformances
-		nestedInterfaceType := checker.Elaboration.InterfaceDeclarationType(nestedInterfaceDeclaration)
-		nestedInterfaceType.ExplicitInterfaceConformances =
-			checker.explicitInterfaceConformances(nestedInterfaceDeclaration, nestedInterfaceType)
+		// Declare nested declarations' members
+		for _, nestedInterfaceDeclaration := range declaration.Members.Interfaces() {
+			// resolve conformances
+			nestedInterfaceType := checker.Elaboration.InterfaceDeclarationType(nestedInterfaceDeclaration)
+			nestedInterfaceType.ExplicitInterfaceConformances =
+				checker.explicitInterfaceConformances(nestedInterfaceDeclaration, nestedInterfaceType)
 
-		checker.declareInterfaceMembers(nestedInterfaceDeclaration)
-	}
+			checker.declareInterfaceMembersAndValue(nestedInterfaceDeclaration)
+		}
 
-	for _, nestedCompositeDeclaration := range declaration.Members.Composites() {
-		checker.declareCompositeLikeMembersAndValue(nestedCompositeDeclaration, ContainerKindInterface)
-	}
+		declareNestedEvent := func(nestedCompositeDeclaration ast.CompositeLikeDeclaration) {
+			checker.declareCompositeLikeMembersAndValue(nestedCompositeDeclaration, ContainerKindComposite)
+
+			// Declare nested composites' values (constructor/instance) as members of the containing composite
+			identifier := *nestedCompositeDeclaration.DeclarationIdentifier()
+
+			// Find the value declaration
+			nestedCompositeDeclarationVariable :=
+				checker.valueActivations.Find(identifier.Identifier)
+
+			eventMembers.Set(
+				nestedCompositeDeclarationVariable.Identifier,
+				&Member{
+					Identifier:            identifier,
+					Access:                checker.accessFromAstAccess(nestedCompositeDeclaration.DeclarationAccess()),
+					ContainerType:         interfaceType,
+					TypeAnnotation:        NewTypeAnnotation(nestedCompositeDeclarationVariable.Type),
+					DeclarationKind:       nestedCompositeDeclarationVariable.DeclarationKind,
+					VariableKind:          ast.VariableKindConstant,
+					ArgumentLabels:        nestedCompositeDeclarationVariable.ArgumentLabels,
+					IgnoreInSerialization: true,
+					DocString:             nestedCompositeDeclaration.DeclarationDocString(),
+				})
```

Comment:
Looks like this function doesn't have to be a closure function. If that's the case, maybe extract out to top level